### PR TITLE
[virt_autotest] ensure default pool size on vm host for vir network test

### PR DESCRIPTION
**Pull request by guoxuguang**
Refer to the current virtual routed network test required, there will be deployed four guest systems(domain) on vm hosts. So, there was required the enough available size of default pool  - default|images on vm hosts. 

So, ensure the enough available default pool size before virtual routed network test as this PR purpose. 

- Verification run: 
[gi-guest_developing-on-host_sles12sp5-kvm](http://149.44.176.58/tests/5058473)
[gi-guest_developing-on-host_sles12sp5-xen](http://149.44.176.58/tests/5060792)
[gi-guest_developing-on-host_sles15sp2-kvm](http://149.44.176.58/tests/5061443)
[gi-guest_developing-on-host_sles15sp2-xen ](http://149.44.176.58/tests/5061442)

### CODE REVIEW

File: lib/virt_autotest/virtual_network_utils.pm [hunk 1]
```diff
@@ -41,7 +41,14 @@ use virt_utils;
 
 our @EXPORT
   = qw(download_network_cfg prepare_network restore_standalone destroy_standalone restart_network restore_guests restore_network
-  destroy_vir_network restore_libvirt_default enable_libvirt_log upload_debug_log check_guest_status check_guest_module save_guest_ip test_network_interface hosts_backup hosts_restore);
+  destroy_vir_network restore_libvirt_default enable_libvirt_log upload_debug_log check_guest_status check_guest_module save_guest_ip test_network_interface hosts_backup hosts_restore get_available_pool);
+
+sub get_available_pool {
```
Comment (pdostal) [inline lib/virt_autotest/virtual_network_utils.pm:46]:
```suggestion
sub get_pool_available_space {
```

File: tests/virt_autotest/libvirt_routed_virtual_network.pm [hunk 2]
```diff
@@ -35,6 +35,11 @@ use version_utils 'is_sle';
 sub run_test {
     my ($self) = @_;
 
+    #Ensure that there is enough available default pool size for virtual network test
+    my $POOL = $self->get_available_pool();
+    die "Available POOL=$POOL. The SUT needs at least 80G of POOL" if $POOL < 80;
```
Comment (pdostal) [inline tests/virt_autotest/libvirt_routed_virtual_network.pm:40]: In here guests are already created. I would execute this where you create your guests.

File: lib/virt_autotest/virtual_network_utils.pm [hunk 3]
```diff
@@ -41,7 +41,14 @@ use virt_utils;
...
+    if (is_xen_host()) {
+        my $pool =  script_output q@virsh pool-info default | grep ^Available | awk '{print $2}'@;
```
Comment (pdostal) [inline lib/virt_autotest/virtual_network_utils.pm:48]: I would add
```perl
script_run 'df -h';
```
just to have some debug output.

File: tests/virt_autotest/libvirt_routed_virtual_network.pm [hunk 4]
```diff
@@ -35,6 +35,11 @@ use version_utils 'is_sle';
 sub run_test {
     my ($self) = @_;
 
+    #Ensure that there is enough available default pool size for virtual network test
+    my $POOL = $self->get_available_pool();
+    die "Available POOL=$POOL. The SUT needs at least 60G of POOL on the xen host for the virtual network" if $POOL < 60;
```
Comment (waynechen55) [inline tests/virt_autotest/libvirt_routed_virtual_network.pm:40]: Suggest use record_info("Detect pool size"); assert_script_run("test $pool >= 60G"); to record test results instead of using Perl die directly. record_info and assert_script_run are test API to be employed to record your test.

Comment (waynechen55) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm:40]: You should append "GiB" to pool size in your message output

Comment (pdostal) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm:40]: I agree to add `record_info("Detect pool size: $POOL");` but `assert_script_run("test $pool >= 60G");` is just wasting time and Perl conditions as well as `die()` are OK. This is minor issue tho.

Comment (waynechen55) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm:40]: Generally speaking, die is not a very good choice for test case checkpoint writing from my POV. You can put die in a function that accepts wrongly arguments, but I think assert_script_run is better for checkpoint. And assert_script_run also conforms to the style in this ibvirt_routed_virtual_network.pm.

File: lib/virt_autotest/virtual_network_utils.pm (hunk 1 above)
Comment (guoxuguang) [reply lib/virt_autotest/virtual_network_utils.pm]: DONE

File: tests/virt_autotest/libvirt_routed_virtual_network.pm (hunk 4 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm:41]: Had added `record_info()`

Comment (pdostal) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm:41]: > Generally speaking, die is not a very good choice for test case checkpoint writing from my POV. You can put die in a function that accepts wrongly arguments, but I think assert_script_run is better for checkpoint. And assert_script_run also conforms to the style in this ibvirt_routed_virtual_network.pm.

Both `assert_script_run()` and `die()` will die. But as I said, this is minor, I'm OK with Wayne's solution too.

File: lib/virt_autotest/virtual_network_utils.pm [hunk 5]
```diff
@@ -41,7 +41,14 @@ use virt_utils;
 
 our @EXPORT
   = qw(download_network_cfg prepare_network restore_standalone destroy_standalone restart_network restore_guests restore_network
-  destroy_vir_network restore_libvirt_default enable_libvirt_log upload_debug_log check_guest_status check_guest_module save_guest_ip test_network_interface hosts_backup hosts_restore);
+  destroy_vir_network restore_libvirt_default enable_libvirt_log upload_debug_log check_guest_status check_guest_module save_guest_ip test_network_interface hosts_backup hosts_restore get_pool_available_space);
+
+sub get_pool_available_space {
+    if (is_xen_host()) {
```
Comment (pdostal) [inline lib/virt_autotest/virtual_network_utils.pm:47]:
```bash
# virsh pool-info default | grep ^Available | awk '{print $2}'
183.65
```
This works on KVM too.

File: lib/virt_autotest/virtual_network_utils.pm (hunk 3 above)
Comment (guoxuguang) [reply lib/virt_autotest/virtual_network_utils.pm]: Great, agree on this. Had added some debug output about hard disk info

Comment (guoxuguang) [reply lib/virt_autotest/virtual_network_utils.pm]: DONE

File: tests/virt_autotest/libvirt_routed_virtual_network.pm (hunk 2 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: sure, had added the comments as your mentioned.

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: DONE

File: tests/virt_autotest/libvirt_routed_virtual_network.pm (hunk 4 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: great, @pdostal @waynechen55 thanks so much for your great help and suggestion. 
We had the same point here. updated to use with `assert_script_run()` and `record_info()` to call out OSD API.

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: DONE

File: tests/virt_autotest/libvirt_routed_virtual_network.pm [hunk 6]
```diff
@@ -35,6 +35,13 @@ use version_utils 'is_sle';
 sub run_test {
     my ($self) = @_;
 
+    #Deploy guest system under default pool - /var/lib/libvirt as default on vm host
```
Comment (pdostal) [inline tests/virt_autotest/libvirt_routed_virtual_network.pm:38]: This shouldn't be here as it applies for all the networking cases￼. otherwise this is ready to go.￼

**Review by pdostal — CHANGES_REQUESTED** [review]

File: tests/virt_autotest/libvirt_routed_virtual_network.pm (hunk 6 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: Great, got it. make make sense here. 
Thanks so much for your comments here.
Had removed it already.

File: lib/virt_autotest/virtual_network_utils.pm (hunk 1 above)
Comment (pdostal) [reply lib/virt_autotest/virtual_network_utils.pm]: I think this should be in `libvirt_virtual_network_init.pm`.

File: tests/virt_autotest/libvirt_routed_virtual_network.pm (hunk 2 above)
Comment (pdostal) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: I think this should be in libvirt_virtual_network_init.pm.

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: sure, had moved to libvirt_virtual_network_init.pm

File: lib/virt_autotest/virtual_network_utils.pm (hunk 5 above)
Comment (guoxuguang) [reply lib/virt_autotest/virtual_network_utils.pm]: Good point! should be coved this both on KVM and XEN together. 
Support to ensure available disk space for default(images) pool both kvm and xen together.

File: tests/virt_autotest/libvirt_routed_virtual_network.pm (hunk 2 above)
Comment (pdostal) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: Thank you 😄

**Review by pdostal — APPROVED** [review]
From my point of view this is good to go 🚀

File: tests/virt_autotest/libvirt_virtual_network_init.pm [hunk 7]
```diff
@@ -45,6 +45,12 @@ sub run_test {
         assert_script_run("test $MEM -ge 20", fail_message => "The SUT needs at least 20G FREE MEM for virtual network test");
     }
 
+    #After deployed guest systems, ensure default pool have at least
+    #60Gib available disk space on vm host for virtual network test
+    my $POOL = $self->get_pool_available_space();
+    record_info('Detect Available POOL', $POOL . 'Gib');
+    assert_script_run("test $POOL -ge 60", fail_message => "The SUT needs at least 60Gib available space of default pool for virtual network test");
```
Comment (grisu48) [inline tests/virt_autotest/libvirt_virtual_network_init.pm:52]: Since `$pool` is available as an internal variable here, why not move the step away from the openQA virtual machine and do something like

    die("The SUT needs at least 60Gib available space of default pool for virtual network test") if ($pool < 60);

instead of the rather slow `assert_script_run`?

**Review by grisu48 — CHANGES_REQUESTED** [review]
Very good work, thank you for your PR! Found one minor issue and would like to hear your opinion on my suggestion.

File: tests/virt_autotest/libvirt_virtual_network_init.pm (hunk 7 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm:52]: @grisu48 thanks so much for your comment. 
we had discuss about using assert_script_run() and die() already.

At last, we had agreed on  using our OSD API - assert_script_run() here, instead of die().  

Thus, there was used assert_script_run().

Comment (grisu48) [reply tests/virt_autotest/libvirt_virtual_network_init.pm:52]: Ok I see. I don't fully get the advantage of using `assert_script_run` here, but consider this a minor issue so feel free to proceed.

**Review by grisu48 — APPROVED** [review]

File: tests/virt_autotest/libvirt_routed_virtual_network.pm (hunk 4 above)
Comment (grisu48) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: > Generally speaking, die is not a very good choice for test case checkpoint writing from my POV. You can put die in a function that accepts wrongly arguments, but I think assert_script_run is better for checkpoint. And assert_script_run also conforms to the style in this ibvirt_routed_virtual_network.pm.

I don't get this argument. `assert_script_run` uses [croak](https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L958) which is equivalent to `die`, so I don't see where is the advantage of using `assert_script_run` over `die` for such a simple comparison. What am I missing?

Comment (grisu48) [reply tests/virt_autotest/libvirt_routed_virtual_network.pm]: I don't want to block this PR, it's OK to go as this is a minor issue. I'm only wondering why `assert_script_run` is considered a better approach than `die` here. I'm just curious :-)

**Comment by guoxuguang** [comment]
Add the latest validation run:
[gi-guest_developing-on-host_sles12sp5-kvm](http://149.44.176.58/tests/5058473)
[gi-guest_developing-on-host_sles12sp5-xen](http://149.44.176.58/tests/5060792)
[gi-guest_developing-on-host_sles15sp2-kvm](http://149.44.176.58/tests/5061443)
[gi-guest_developing-on-host_sles15sp2-xen ](http://149.44.176.58/tests/5061442)

@alice-suse @Julie-CAO @waynechen55 @pdostal @grisu48 please help me review this PR, if have time.

File: lib/virt_feature_test_base.pm [hunk 8]
```diff
@@ -179,4 +180,17 @@ sub get_free_mem {
     }
 }
 
+sub get_pool_available_space {
```
Comment (alice-suse) [inline lib/virt_feature_test_base.pm:183]: Better to name it get_active_pool_and_available_space, and return both active_pool_name and available size.

File: tests/virt_autotest/libvirt_virtual_network_init.pm [hunk 9]
```diff
@@ -45,6 +45,12 @@ sub run_test {
         assert_script_run("test $MEM -ge 20", fail_message => "The SUT needs at least 20G FREE MEM for virtual network test");
     }
 
+    #After deployed guest systems, ensure default pool have at least
+    #60Gib available disk space on vm host for virtual network test
+    my $POOL = $self->get_pool_available_space();
+    record_info('Detect Available POOL', $POOL . 'Gib');
```
Comment (alice-suse) [inline tests/virt_autotest/libvirt_virtual_network_init.pm:51]: And record the info of active_pool_name and available space size here, which is useful for debug.

File: tests/virt_autotest/libvirt_virtual_network_init.pm (hunk 7 above)
Comment (alice-suse) [inline tests/virt_autotest/libvirt_virtual_network_init.pm:52]: suggest "of default pool" -> "of active pool"

**Comment by alice-suse** [comment]
@guoxuguang Besides the active pool size check and assertion, I am afraid this PR misses a very important part that finding a size large enough partition to be used as active pool directory( libvirt/image) if the used one size is small.

**Comment by guoxuguang** [comment]
> @guoxuguang Thanks for the PR! Besides the active pool size check and assertion, I am afraid this PR misses a very important part that finding a size large enough partition to be used as active pool directory( libvirt/image) if the used one size is small.
sure, I will cre

File: lib/virt_feature_test_base.pm (hunk 8 above)
Comment (guoxuguang) [reply lib/virt_feature_test_base.pm]: sure, had changed as `get_active_pool_and_available_space(),` and return both `$active_pool_name` and `$available_size` now.

File: tests/virt_autotest/libvirt_virtual_network_init.pm (hunk 9 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm]: Great idea. had added record the info of active_pool_name and available space size  together.

File: tests/virt_autotest/libvirt_virtual_network_init.pm (hunk 7 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm]: Great, had updated here already.

**Comment by guoxuguang** [comment]
> @guoxuguang Thanks for the PR! Besides the active pool size check and assertion, I am afraid this PR misses a very important part that finding a size large enough partition to be used as active pool directory( libvirt/image) if the used one size is small.

Good point, check and ensure active_pool_name and available_pool_size for this PR now.
sure, as your mentioned about finding a size large enough partition to be used as active pool directory( libvirt/image) if the used one size is small. I will create a new PR for this.

**Comment by guoxuguang** [comment]
Add the latest Validation run:
wip [gi-guest_developing-on-host_sles12sp5-kvm](http://149.44.176.58/tests/5072827#)

**Comment by Julie-CAO** [comment]
LGTM

**Comment by guoxuguang** [comment]
Quick fix, refer to the latest validation run: [gi-guest_developing-on-host_sles12sp5-kvm ](https://149.44.176.58/tests/5072827) ran on the testing machine - grenache-1:10 have the large Hard Disk - 5TiB+. 
latest code changed to support this PR ran on the testing machine have Hard Disk TiB or GiB. 
Add the latest validation run:
[5TiB SUT](http://149.44.176.58/tests/5078281)
[300GiB SUT](http://149.44.176.58/tests/5078280)

**Comment by pdostal** [comment]
Revealing and covering those corner cases was pretty neat, @guoxuguang ! 👍

**Comment by guoxuguang** [comment]
> Revealing and covering those corner cases was pretty neat, @guoxuguang ! +1

Cheers!
